File: books/introduction_to_design_with_good_code_bad_code/3_class_design/money.go
package money

import "fmt"

type Money struct {
	Amount   int
	Currency string
}

func New(amount int, currency string) (*Money, error) {
	if amount < 0 {
		return nil, fmt.Errorf("Argument Error: specify amount more than 0")
	}

	if len(currency) == 0 {
		return nil, fmt.Errorf("Argument Error: specify currency")
	}

	return &Money{
		Amount:   amount,
		Currency: currency,
	}, nil
}

func (m Money) Add(other Money) (*Money, error) {
	if m.Currency != other.Currency {
		return nil, fmt.Errorf("Argument Error: difference currency")
	}
	added := m.Amount + other.Amount
	return New(added, m.Currency)
}

File: books/introduction_to_design_with_good_code_bad_code/3_class_design/money.rb
class Money
  attr_reader :amount, :currency

  def initialize(amount, currency)
    if amount < 0
      raise ArgumentError.new("specify amount more than 0.")
    end
    if currency.nil? || currency.empty?
      raise ArgumentError.new("specify currency.")
    end

    @amount = amount
    @currency = currency
    self.freeze
  end

  def add(other)
    if @currency != other.currency
      raise ArgumentError.new("different currency.")
    end
    added = @amount + other.amount
    return Money.new(added, @currency)
  end
end

one_handread = Money.new(100, "en")
two_handread = one_handread.add(Money.new(100, "en"))
puts one_handread.amount, one_handread.currency
puts two_handread.amount, two_handread.currency
